Add support for instructions with +rb +rw +rd +ro 8bit encoding

xoofx opened the pull request:
Hey, so this PR is prototyping the changes talked in issue #16

This is not really intended to be merged but more as a base for discussion about how the API could evolved to allow easier encoding of instructions.

## Naming

I have replaced naming like `Code.Mov_Eq_Gq` to `Code.Mov_rm64_r64`. The rationale behind this is to be able to read an instruction manual and figure out immediately which instruction to use in Iced (In the instruction manual whenever there is a space or `,` I replaced it by `_`, and the characters `*` or `/` are removed.

I found the actual notation quite unfamiliar to work with (`Eq_Gq`) and quite difficult to read

The same would apply to other instructions like for push:

![image](https://user-images.githubusercontent.com/715038/45584310-7bd48100-b8d1-11e8-91a8-cf58500e44ac.png)

We would have things like:

* `PUSH_rm16`, `PUSH_rm32`, `PUSH_rm64`
* `PUSH_r16`, `PUSH_r32`, `PUSH_r64`
* `PUSH_imm8`, `PUSH_imm16`, `PUSH_imm32`

Pondering between`PUSH` or `Push` which one would be better actually. In this PR I have kept the name `Mov` but not sure...

## Handling of `+rb` `+rw` `+rd` encodings

I'm not particularly convinced by the changes I did. I have slightly modified the `OpCodeHandlers` to handle the following instructions:

`Code.Mov_r8_imm8`, `Code.Mov_r16_imm16`, `Code.Mov_r32_imm32`, `Code.Mov_r64_imm64`, and introduced subclasses of `LegacyHandler` like `LegacyHandler_r8`, `LegacyHandler_r16`...etc.

It is more a hack than a deep thought about encoding. I'm wondering if it is worth to have a prepass of `var info = new uint[...]` descriptions to convert them to the proper instance rather than having this directly instantiated in the array (we would avoid the loop at the end, but the managed assembly is actually more optimized with a table of uint[] and a code to instantiate instances, the JIT would likely suffer more from running a handwritten table of managed instances)

In some way, I would think that  the encoding tables found in `OpCodeHandlers` would be in the end closer to the kind of tables we have for decoding like in `OpCodeHandlers32Tables`

There are a couple of tests failing that I haven't checked deeply. Likely something that I simplified in this PR that would require a few more stuffs. 

But anyway, it is mostly for you to get a feeling about the changes. You would likely want something very different!

xoofx commented:
> This doesn't show that it's a sign-extended 32-bit immediate. The Intel manual also uses imm32 in this case. imm32to64 shows that it's sign extended to 64 bits (there are also imm8 -> imm64 immediates).

good catch, that's was a typing error. Not sure we need to explicit the imm32to64, the mnemonic in the intel doesn't. That would be more a comment. In that case `MOV_rm64_imm32` looks fine to me (I can match it quickly with Intel doc)

> What about the 3rd operand: VPSHUFB xmm1 {k1}{z}, xmm2, xmm3/m128, xmmm128 looks weird to me.

Yeah, I haven't thought about the other instructions. For these, their notation would be encoded differently though no? something like `PSHUFB_SSSE3` or should we use instead a notation similar to the one used by C/C++ intrinsics? (like [here](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=PSHUFB))
I don't know. 

> It's optimized for code size, and it's possible that it's faster too (I didn't check). The array is initialized from data in the PE file so what is thousands of lines of data in the source code is just a few IL instructions.

Yeah, that's what I amended in the previous post. The IL generated arrays with direct managed instances would be slower to initialize actually, so likely better to keep the uint[] approach.

> Either is fine with me. I picked Push because it better matches the .NET naming convention.

Same here, I don't mind. As there was already `_` the naming convention is already a bit twisted so we can probably choose whatever.

0xd4d commented:
> Not sure we need to explicit the imm32to64, the mnemonic in the intel doesn't

There are some instructions that have both an Ib and Id operand, eg. add rm64, Ib/Id, `81 /0` and `83 /0`. In those cases it's required. And for consistency it should be added to the other ones too.

> Yeah, I haven't thought about the other instructions. For these, their notation would be encoded differently though no? something like PSHUFB_SSSE3 or should we use instead a notation similar to the one used by C/C++ intrinsics?

IMHO, the CPUID feature bit shouldn't be part of the Code value. The encoding is part of the name because of a few duplicate instructions and so it's obvious which encoding will be used.

xmm/m128: xm128, XMMm128, xmmM128, xmmm128

0xd4d commented:
I know Intel uses 'imm8' in those two examples above (81/0 and 83/0) but IMHO it's better to show that the value is actually sign extended to the correct size, in this case 64 bits.

0xd4d commented:
I also don't know if it's useful to include the memory size in the Code name. It's only needed in a few instructions eg. fldenv. Many EVEX instructions can broadcast a value so the memory operand can be 128 bits or 32 bits etc. (`VUNPCKLPS xmm1 {k1}{z}, xmm2, xmm3/m128/m32bcst`)

0xd4d commented:
Perhaps: `m128b32` or ignore the broadcast size: `m128`

0xd4d commented:
I've created a PR https://github.com/0xd4d/iced/pull/18 that only renames the Code values. It doesn't merge the Code values that encode the register bits in the low 3 bits of the opcode.

xoofx commented:
@0xd4d so, I made a prototype (and removed previous one) for `Inc_r16`/`Inc_r32` based on your suggestion. I have added a Operand encoder `OpRegOpCodeByte` (not great, If you have a better name, I'm out of ideas...). 
Let me know if it is the way you wanted to do this change.

0xd4d commented on Iced/Intel/EncoderInternal/OpHandlers.cs:
```diff
@@ -370,6 +373,9 @@ static class LegacyOps {
 			new OpRegSTi(),
+			new OpRegOpCodeByte(Register.AX, Register.DI),
+			new OpRegOpCodeByte(Register.EAX, Register.EDI),
+			new OpRegOpCodeByte(Register.RAX, Register.RDI),
```
You'll probably need three more handlers for the other instructions since they can use all 16 regs.

0xd4d replied:
No, I realized you don't need it. The encoder verifies that only the low 8 regs are used if it's 16 or 32-bit mode, so you can change the last reg to R15W/R15D/R15 here.

0xd4d commented on Iced/Intel/EncoderInternal/OpHandlers.cs:
```diff
@@ -629,6 +635,26 @@ sealed class OpModRM_rm : Op {
+
+		public override void Encode(Encoder encoder, ref Instruction instr, int operand) 
+		{
+			if (instr.GetOpKind(operand) == OpKind.Register) 
```
You should call encoder.Verify() to verify that it's a register OpKind.

nit: I put braces on the same line as the 'if' and the method decl.

0xd4d commented on Iced/Intel/EncoderInternal/OpHandlers.cs:
```diff
@@ -629,6 +635,26 @@ sealed class OpModRM_rm : Op {
+			{
+				var register = instr.GetOpRegister(operand);
+				encoder.Verify(operand, register, regLo, regHi);
+				encoder.OpCode = encoder.OpCode + (uint)(register - regLo);
```
This works if only low 8 registers are used. If this handler gets reused for all 16 regs, you'll need a `&7` here too

0xd4d replied:
Also if it's reg r8-r15, you need to set encoder.Flags' REX.W bit. There's also the 8bit regs that need special handling. The encoder already has some logic in `AddModRMRegister`, but it's for modrm instructions, you could add a new method.

xoofx replied:
So just to make sure, I would add a new `OpRegEmbed32`  that would take two ranges of register RAX-RDI and R8W-R15W, and I would detect either R8W-R15W  usage to add proper REX bits, that's correct?

What about the name `OpRegOpCodeByte`, should I use something like  `OpRegEmbed8` instead?

xoofx replied:
Ok, though for the following case:

```c#
				(uint)EncFlags1.Legacy | (uint)Code.Push_AX | ((uint)0x50 << (int)EncFlags1.OpCodeShift), (uint)LegacyFlags.Encodable_Any | (uint)LegacyFlags.o16, (uint)LegacyOpKind.AX,
				(uint)EncFlags1.Legacy | (uint)Code.Push_R8W | ((uint)0x50 << (int)EncFlags1.OpCodeShift), (uint)LegacyFlags.Encodable_Only64 | (uint)LegacyFlags.o16_rexb, (uint)LegacyOpKind.R8W,
				(uint)EncFlags1.Legacy | (uint)Code.Push_EAX | ((uint)0x50 << (int)EncFlags1.OpCodeShift), (uint)LegacyFlags.Encodable_Only1632 | (uint)LegacyFlags.o32, (uint)LegacyOpKind.EAX,
				(uint)EncFlags1.Legacy | (uint)Code.Push_RAX | ((uint)0x50 << (int)EncFlags1.OpCodeShift), (uint)LegacyFlags.Encodable_Only64, (uint)LegacyOpKind.RAX,
				(uint)EncFlags1.Legacy | (uint)Code.Push_R8 | ((uint)0x50 << (int)EncFlags1.OpCodeShift), (uint)LegacyFlags.Encodable_Only64 | (uint)LegacyFlags.rexb, (uint)LegacyOpKind.R8,
```

How can I put a single line for `Push_r16` while the Legacy flags can be `LegacyFlags.Encodable_Any | (uint)LegacyFlags.o16,` for AX-DI and `LegacyFlags.Encodable_Only64 | (uint)LegacyFlags.o16_rexb` for R8W-R15W?

xoofx replied:
Or should I add for them a `Push_r16_rexb`?

0xd4d replied:
You can remove the rexb part. The handler should set it if it's a high reg (or better, a new Encoder method called by the handler that has all the logic). In other words, the lines should look very similar to the Inc_xxx lines. Encodable_Any should be used in this case though since the instructions are available in all modes and not only 16/32-bit modes (inc/dec r16/32)

> Or should I add for them a Push_r16_rexb?

No, Push_r16 can support all 16 regs.

xoofx replied:
yep ok, will try to do that then

xoofx commented:
question: When encoding an instruction, does the Instruction.CodeSize has to be checked? Getting some exceptions with my changes on `EncoderTest_Invalid` with `push_r16`/`pop_r16` likely something wrong somewhere but I'm not sure.

For example, for the following input
`codeSize: 16, code: Pop_r16, hexBytes: "59", invalidCodeSize: 32`
the original instruction decoded is `pop cx`

and the encoded is running successfully, but the tests is expecting an encoding error... or is it the `EncodeInvalid_Data` with `DecoderTestUtils.GetDecoderTests` that is not working with my changes... wondering... any ideas?

xoofx commented:
Oh, ok, it was coming from `DecoderTestUtils` in GetCode64(), as previously it was checking for 
`Push_R8W`.. `Push.R16W`, but actually they are now not completely valid for Push_r16... (only some registers are not).. not sure how we will handle this invalid opcodes (would require to pass register informations...)

0xd4d commented:
You can remove it from the Code64 property, it's supposed to be for instructions that are 100% 64-bit only. The old code value was eg. Push_R14W, which was 64-bit only.

0xd4d commented on Iced.UnitTests/Intel/InstructionTests/InstructionTests_Misc.cs:
```diff
@@ -54,7 +54,7 @@ public sealed class InstructionTests_Misc {
 			int maxValue = GetEnumSize(typeof(Code)) - 1;
 			Assert.True(maxValue < (1 << Instruction.TEST_CodeBits));
-			Assert.True(maxValue >= (1 << (Instruction.TEST_CodeBits - 1)));
+			Assert.True(maxValue >= (1 << (Instruction.TEST_CodeBits - 2)));
```
You should update CodeBits in Instruction.CodeFlags and its comment.

xoofx replied:
Originally, I did that, but it was weird to have a hole with `RoundingControlShift` (that I didn't change), if that's ok, I can change `CodeFlags.CodeBits` only otherwise do you want to shift everything? (Not sure about the implications)

0xd4d replied:
Just the bit count is enough. It will regain its extra bit in the near future when I add undocumented instructions and Intel's deprecated KNC instructions.

0xd4d commented on Iced/Intel/DecoderInternal/OpCodeHandlers64.cs:
```diff
@@ -2012,33 +2012,33 @@ sealed class OpCodeHandler_RegIb3 : OpCodeHandler {
 		}
 
 		static readonly (Code code, Register register)[] noRexPrefix = new(Code code, Register register)[8] {
-			(Code.Mov_AL_imm8, Register.AL),
```
The Code value is, after the changes, always identical here and can be removed (same thing with info16/32/64 tables below)

0xd4d commented on Iced/Intel/Encoder.cs:
```diff
@@ -632,6 +632,63 @@ public sealed class Encoder {
+			}
+
+			var reg = instr.GetOpRegister(operand);
+			uint regNum = (uint)(reg - regLo);
```
Can the logic below here be simplified. Have a look at AddModRMRegister, it can handle all GPR8, GPR16, GPR32, GPR64 and contains only 2 if statements.

0xd4d commented on Iced/Intel/HexUtils.cs:
```diff
@@ -52,6 +53,21 @@ static class HexUtils {
+
+			var builder = new StringBuilder(hexData.Length * 3 - 1);
+			for(int i = 0; i < hexData.Length; i++) {
+				builder.Append(hexData[i].ToString("2X"));
```
Should this be "X2"?

xoofx commented:
I have pushed all fixes. Let me know if there is anything remaining.

0xd4d commented on Iced.UnitTests/Intel/EncoderTests/EncoderTest.cs:
```diff
@@ -108,9 +108,9 @@ public abstract class EncoderTest {
 			var origInstrCopy = origInstr;
 			bool result = encoder.TryEncode(ref origInstr, origRip, out int encodedInstrLen, out string errorMessage);
-			Assert.True(errorMessage == Encoder.ERROR_ONLY_1632_BIT_MODE || errorMessage == Encoder.ERROR_ONLY_64_BIT_MODE);
+			Assert.True(errorMessage == Encoder.ERROR_ONLY_1632_BIT_MODE || errorMessage == Encoder.ERROR_ONLY_64_BIT_MODE, "Unexpected error Message: " + errorMessage ?? "null");
```
Parens are needed: `(errorMessage ?? "null")`

0xd4d commented on Iced/Intel/Code.cs:
```diff
@@ -445,53 +343,11 @@ public enum Code {
 
 		Nopw,												// o16 90
-		Xchg_R8W_AX,										// o16 REX.B 90
+		Xchg_AX_r16,										// o16 REX.B 90
```
Xchg must be renamed: ax/eax/rax is the 2nd operand.

0xd4d commented:
AddReg still looks too big. eg. it adds P66, but is that needed? The encoder table should already have the P66 flag (as 'o16').

0xd4d commented:
But this isn't important to fix, as long as the tests pass. I can have a look at it more some other day, unless you want to check if it's possible to remove some of the code.

xoofx commented:
> AddReg still looks too big. eg. it adds P66, but is that needed? The encoder table should already have the P66 flag (as 'o16').

Indeed, this should be fixed by latest commit

0xd4d commented:
Looks very good, if the tests pass I'll merge it, unless you have more commits waiting!

xoofx commented:
> Looks very good, if the tests pass I'll merge it, unless you have more commits waiting!

Tests are passing, hopefully I think I'm done 😅 👍

0xd4d commented:
Thanks! Merged.
